src/ui/public/accessibility/kbn_accessible_click.js

```diff
@@ -0,0 +1,86 @@
+ *
+ * If you need to use a <div>, <p>, or <a> without the href attribute, then you need to allow
+ * them to receive focus and to respond to keyboard input. The workaround is to:
+ *
+ *   - Give the element tabindex="0" so that it can receive keyboard focus.
+ *   - Add a JavaScript onkeyup event handler that triggers element functionality if the Enter key
+ *     is pressed while the element is focused. This is necessary because some browsers do not trigger
+ *    onclick events for such elements when activated via the keyboard.
+ *   - If the item is meant to function as a button, the onkeyup event handler should also detect the
+ *     Spacebar in addition to the Enter key, and the element should be given role="button".
+ *
+ * Apply this directive to any of these elements to automatically do the above.
+ */
+
+import { accessibleClickKeys } from './accessible_click_keys';
+import { uiModules } from 'ui/modules';
+
+uiModules.get('kibana')
+.directive('kbnAccessibleClick', function ($parse) {
+  return {
+    restrict: 'A',
+    scope: {
+      kbnAccessibleClick: '&',
+    },
+    controller: $element => {
+      $element.on('keydown', e => {
+        // We don't care if the user is actually interacting with a different element.
+        if (e.currentTarget !== e.target) {
+          return;
+        }
+
+        // Prevent a scroll from occurring if the user has hit space.
+        if (accessibleClickKeys[e.keyCode]) {
+          e.preventDefault();
+        }
+      });
+    },
+    link: (scope, element, attrs) => {
+      // The whole point of this directive is to hack in functionality that native buttons provide
+      // by default.
+      const elementType = element.prop('tagName');
+
+      if (elementType === 'BUTTON') {
+        throw new Error(`kbnAccessibleClick doesn't need to be used a button.`);
+      }
+
+      if (elementType === 'A' && attrs.href !== undefined) {
+        throw new Error(`kbnAccessibleClick doesn't need to be used on a link if it has a href attribute.`);
+      }
+
+      // If the developer hasn't already specified attributes required for accessibility, add them.
+      if (attrs.tabindex === undefined) {
+        element.attr('tabindex', '0');
+      }
+
+      if (attrs.role === undefined) {
+        element.attr('role', 'button');
+      }
+
+      element.on('keyup', e => {
+        // Support keyboard accessibility by toggling display on ENTER or SPACE keypress.
+        if (accessibleClickKeys[e.keyCode]) {
+          // If a unique click handler has been specified, then call it.
+          if (attrs.kbnAccessibleClick) {
+            return scope.kbnAccessibleClick();
+          }
+
+          // Otherwise, default to the specified ng-click.
+          const ngClick = $parse(attrs.ngClick);
+
+          if (!ngClick) {
+            throw new Error('kbnAccessibleClick requires a callback or ng-click defined on its element.');
+          }
+
+          ngClick(scope.$parent);
```

Comment:
I hadn't considered using this directive in conjunction with ng-click, it's a great idea! You've got me wondering, do we even want to provide the ability to specify a separate event handler on this directive, or should we always require ng-click? My gut says the latter because it would ensure the interaction is always the same regardless of whether you're using a keyboard or mouse to interact with the element.

I also think we could avoid manually parsing the ng-click attribute if we simply fire a click event and allow ng-click to do its work as usual. This block could be simplified to:

```
      element.on('keyup', e => {
        if (accessibleClickKeys[e.keyCode]) {
          element.click();
        }
      });
```

One other question along this line of thought: you mentioned in https://github.com/elastic/kibana/issues/11513 that listening to the keyup event was necessary because "some browsers do not trigger onclick events for such elements when activated via the keyboard." Are there browsers that *do* trigger click events and if so will having an `ng-click` and an `kbn-accessible-click` on the same element cause the event handler to execute twice?